feat(core): map for property to htmlFor

pkozlowski-opensource opened the pull request:
**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x")

```
[x] Feature
```

**What is the current behavior?** (You can also link to an open issue here)

Can't write `<label [for]="ctxProp">`. People need to use `<label [htmlFor]="ctxProp">` or `<label [attr.for]="ctxProp">` which is a common source of confusion.

**What is the new behavior?**

People can write `<label [for]="ctxProp">` and it is equivalent to `<label [htmlFor]="ctxProp">`.

**Does this PR introduce a breaking change?** (check one with "x")

```
[x] No
```

**Other information**:

Changes in `template_parser.ts` are due to clang-format run.

This improves ergonomics a bit by allowing people to write:
`<label [for]="ctxProp"></label>`.
This is similar to the exisiting class -> className mapping.

Closes #7516

vicb commented:
Could this lead to subtile bugs ? ie `<my-list [for]="group">` ?

pkozlowski-opensource commented:
> Could this lead to subtile bugs ? ie <my-list [for]="group"> ?

Nope, apparently not. As a prove see what happens with `readonly` property for which we've got re-mapping in place: http://plnkr.co/edit/VuzJgtMDgGlUFdxC6Y4F?p=preview

In any case we would already have a problem for props like `class`, `readonly` etc. - basically for all re-mapped props.

I didn't dig into the code, but I guess that binding to directive properties goes through different code path as compared to binding to element properties. We can add an integration test to the case you are raising.

vicb commented:
> I didn't dig into the code, but I guess that binding to directive properties goes through different code path as compared to binding to element properties. We can add an integration test to the case you are raising.

The problem is how do you know if you're binding to a directive or an element here ??

vicb commented:
One way to fix would be to error when the binding conflict in directive and this mapping.

pkozlowski-opensource commented:
> The problem is how do you know if you're binding to a directive or an element here ??
> One way to fix would be to error when the binding conflict in directive and this mapping.

@vicb this is well defined already in the compiler. If there is the same property on a directive and the host element the directive "wins" (you can eventually reflect it back on an element manually via `Renderer`) if you wish.

> The problem is how do you know if you're binding to a directive or an element here ??

You first check props of all directives and if this exists it "wins", see above.

So, all in all, I don't think there is any confusion here.

vicb commented:
`<my-list for="group">`

Ok let's head for a coffee before thinking more about this

pkozlowski-opensource commented:
> `<my-list for="group">`

@vicb this doesn't change anything as `for="group"` is just equivalent to `[for]="'group'"` and all the above rules still apply. We are _still_ binding to a directive property here - what matters is a presence of this property (or nor) on a directive rather than a particular syntax used.

Here is the evidence: http://plnkr.co/edit/QUmRZshidTbstjbdYIHc?p=preview

> Ok let's head for a coffee before thinking more about this

Sure, enjoy yours :-) But I still think that rules are crystal-clear here. In other case we would have problems with other mappings.

vicb commented:
> for="group" is just equivalent to [for]="'group'"

... with the subtile difference that the for attribute will be remove from the resulting html in the latter case

pkozlowski-opensource commented:
> with the subtile difference that the for attribute will be remove from the resulting html in the latter case

Sure. But this is not related to re-mapping at all, if I understand correctly. It would be more of a concern for allowed names for directive properties (`id` comes to mind immediately).

vicb commented:
IMO the current behavior, ie error:

```
Can't bind to 'for' since it isn't a known property of 'label'. ("
      <h2>Hello {{name}}</h2>
      
      <label [ERROR ->][for]="field.id"></label>
      
    </div>
"): App@4:13
```

is fine.

We can improve it by adding the places where we now that the prop does not exists, ie adding "do you mean [attr.for] ?" in the message above.

mohammedzamakhan commented on modules/@angular/compiler/src/template_parser/template_parser.ts:
```diff
@@ -32,6 +32,7 @@ import {AttrAst, BoundDirectivePropertyAst, BoundElementPropertyAst, BoundEventA
 import {PreparsedElementType, preparseElement} from './template_preparser';
 
 
+
```
No valuable changes in this file, this can be removed from the commits

mhevery commented:
Discussed with @vicb and we decided that we should do this mapping for all HTML attributes whose Element property does not match the name. @pkozlowski-opensource to create a follow up PR to add the rest of the attributes.

pkozlowski-opensource committed:
feat(core): map 'for' attribute to 'htmlFor' property

This improves ergonomics a bit by allowing people to write:
`<label [for]="ctxProp"></label>`.
This is similar to the exisiting class -> className mapping.

Closes #7516
